### ISSUE TITLE
App freezes when invoking Amplify.DataStore.save()

### ISSUE THREAD

**Issue by Zablas** [aws-amplify/amplify-android#706]
I'm using Kotlin and AWS Amplify to make a simple mobile app. I'll begin by
showing what I have setup. My `schema.graphql`:

```graphql
type OfferGQL @model {
   id: ID!
   name: String!
   description: String!
   address: String!
   price: Float!
 }
```

My AWS plugins are setup like this:
```kotlin
Amplify.addPlugin(AWSCognitoAuthPlugin())
Amplify.addPlugin(AWSS3StoragePlugin())
Amplify.addPlugin(AWSDataStorePlugin())
Amplify.configure(applicationContext)
```

This is how I try to save my data to the DataStore:
```kotlin
// Create a new offer object
val newOffer = OfferGQL.builder()
    .name(textFieldOfferName.text.toString())
    .description(textFieldOfferDescription.text.toString())
    .address(textFieldOfferAddress.text.toString())
    .price(textFieldOfferPrice.text.toString().toFloat())
    .build()

// Put it in the DataStore 
Amplify.DataStore.save(newOffer,
    { result ->
        runOnUiThread{Toast.makeText(this,"Offer added successfully", Toast.LENGTH_LONG).show()}
    },
    { error ->
        println(error.message)
        println(error.localizedMessage)
        println(error.cause)
        println(error.recoverySuggestion)
    }
)
```

The problem is that when I invoke `Amplify.DataStore.save(...)`, the entire app
freezes. I followed the documentation to configure everything so I don't really
know why this is happening.

(Also asked this [on StackOverflow](https://stackoverflow.com/q/63300301/695787).)

**Comment by jamesonwilliams** [comment]
Hey @Zablas -- Can you show us the `dependencies` block of your app module's `build.gradle`? What version of Amplify are you using?

I'm guessing you're using 1.1.0, which shipped with this particular defect. We fixed it for 1.1.1, by including this fix: https://github.com/aws-amplify/amplify-android/pull/653.

**Comment by Zablas** [comment]
These are the dependencies:

`dependencies {
    implementation fileTree(dir: 'libs', include: ['*.jar'])
    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:$kotlin_version"
    implementation 'androidx.appcompat:appcompat:1.1.0'
    implementation 'androidx.core:core-ktx:1.3.0'
    implementation 'androidx.constraintlayout:constraintlayout:1.1.3'
    implementation 'com.squareup.picasso:picasso:2.71828'
    implementation 'com.google.android.material:material:1.1.0'
    implementation 'com.beust:klaxon:5.0.1'
    implementation 'com.android.volley:volley:1.1.1'
    testImplementation 'junit:junit:4.12'
    androidTestImplementation 'androidx.test.ext:junit:1.1.1'
    androidTestImplementation 'androidx.test.espresso:espresso-core:3.2.0'`

`    // AWS`
`    implementation 'com.amplifyframework:core:1.1.1'
    implementation 'com.amplifyframework:aws-datastore:1.1.1'
    implementation 'com.amplifyframework:aws-api:1.1.1'
    implementation 'com.amplifyframework:aws-auth-cognito:1.1.1'
    implementation 'com.amplifyframework:aws-storage-s3:1.1.1'
}`

As for Amplify I get that I'm running 4.26.0 after running "amplify -v"

**Comment by Zablas** [comment]
Here are some updates on the issue:
I've tried reconfiguring the API from scratch by removing the old API and adding a new one. It didn't seem to change anything. Then I found issue #595 which suggested reinstalling the app completely which I did. The function didn't freeze anymore but I got an error saying that I have multiple APIs configured. After removing the API from the cloud using AWS console and removing it from amplifyconfiguration.json my can launch but I'm getting this error now:

`E/amplify:aws-datastore: Error encountered in the DataStore.
    AmplifyException {message=Initial cloud sync failed., cause=AmplifyException {message=Failure performing sync query to AppSync., cause=AmplifyException {message=Could not retrieve the response body from the returned JSON, cause=java.io.IOException: Failed to retrieve Cognito User Pools token., recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Check your internet connection.}
        at com.amplifyframework.datastore.syncengine.SyncProcessor.lambda$createHydrationTask$2$SyncProcessor(SyncProcessor.java:122)
        at com.amplifyframework.datastore.syncengine.-$$Lambda$SyncProcessor$jn-OmlATncPMc9HkFSgk_WBpToE.accept(Unknown Source:4)
        at io.reactivex.internal.operators.completable.CompletablePeek$CompletableObserverImplementation.onError(CompletablePeek.java:88)
        at io.reactivex.internal.operators.completable.CompletableAndThenCompletable$SourceObserver.onError(CompletableAndThenCompletable.java:62)
        at io.reactivex.internal.operators.single.SingleFlatMapCompletable$FlatMapCompletableObserver.onError(SingleFlatMapCompletable.java:97)
        at io.reactivex.internal.operators.observable.ObservableFlatMapCompletableCompletable$FlatMapCompletableMainObserver.onError(ObservableFlatMapCompletableCompletable.java:126)
        at io.reactivex.internal.operators.observable.ObservableFlattenIterable$FlattenIterableObserver.onError(ObservableFlattenIterable.java:125)
        at io.reactivex.internal.observers.BasicFuseableObserver.onError(BasicFuseableObserver.java:100)
        at io.reactivex.internal.operators.observable.ObservableToList$ToListObserver.onError(ObservableToList.java:92)
        at io.reactivex.internal.operators.mixed.SingleFlatMapObservable$FlatMapObserver.onError(SingleFlatMapObservable.java:74)
        at io.reactivex.internal.operators.single.SingleDoOnError$DoOnError.onError(SingleDoOnError.java:63)
        at io.reactivex.internal.operators.single.SingleDoOnSuccess$DoOnSuccess.onError(SingleDoOnSuccess.java:65)
        at io.reactivex.internal.operators.single.SingleCreate$Emitter.tryOnError(SingleCreate.java:94)
        at io.reactivex.internal.operators.single.SingleCreate$Emitter.onError(SingleCreate.java:80)
        at com.amplifyframework.datastore.syncengine.-$$Lambda$3TQsNIUgAvjdYqsJiM71fw48re0.accept(Unknown Source:4)
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$sync$1(AppSyncClient.java:97)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$suvEma05s-p8v-66iE4cswLKYac.accept(Unknown Source:4)
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onFailure(AppSyncGraphQLOperation.java:150)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:525)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: AmplifyException {message=Failure performing sync query to AppSync., cause=AmplifyException {message=Could not retrieve the response body from the returned JSON, cause=java.io.IOException: Failed to retrieve Cognito User Pools token., recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$sync$1(AppSyncClient.java:97) 
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$suvEma05s-p8v-66iE4cswLKYac.accept(Unknown Source:4) 
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onFailure(AppSyncGraphQLOperation.java:150) 
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:525) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
     Caused by: AmplifyException {message=Could not retrieve the response body from the returned JSON, cause=java.io.IOException: Failed to retrieve Cognito User Pools token., recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onFailure(AppSyncGraphQLOperation.java:150) 
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:525) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
     Caused by: java.io.IOException: Failed to retrieve Cognito User Pools token.
        at com.amplifyframework.api.aws.sigv4.AppSyncSigV4SignerInterceptor.intercept(AppSyncSigV4SignerInterceptor.java:210)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
        	... 3 more
     Caused by: AmplifyException {message=getTokens does not support retrieving tokens while signed-out, cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.api.aws.sigv4.DefaultCognitoUserPoolsAuthProvider.fetchToken(DefaultCognitoUserPoolsAuthProvider.java:97)
        at com.amplifyframework.api.aws.sigv4.DefaultCognitoUserPoolsAuthProvider.getLatestAuthToken(DefaultCognitoUserPoolsAuthProvider.java:103)
        at com.amplifyframework.api.aws.sigv4.AppSyncSigV4SignerInterceptor.intercept(AppSyncSigV4SignerInterceptor.java:208)
        	... 6 more
E/amplify:aws-datastore: Failure encountered while attempting to start API sync.
    AmplifyException {message=Initial sync during DataStore initialization failed., cause=AmplifyException {message=Failure performing sync query to AppSync., cause=AmplifyException {message=Could not retrieve the response body from the returned JSON, cause=java.io.IOException: Failed to retrieve Cognito User Pools token., recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=There is a possibility that there is a bug if this error persists. Please take a look at 
    https://github.com/aws-amplify/amplify-android/issues to see if there are any existing issues that 
    match your scenario, and file an issue with the details of the bug if there isn't.}
        at com.amplifyframework.datastore.syncengine.Orchestrator.lambda$startApiSync$4$Orchestrator(Orchestrator.java:271)
        at com.amplifyframework.datastore.syncengine.-$$Lambda$Orchestrator$kWVR_9CuT_WgT9hMLwzl4h2c2Lg.subscribe(Unknown Source:2)
        at io.reactivex.internal.operators.completable.CompletableCreate.subscribeActual(CompletableCreate.java:39)
        at io.reactivex.Completable.subscribe(Completable.java:2309)
        at io.reactivex.internal.operators.completable.CompletablePeek.subscribeActual(CompletablePeek.java:51)
        at io.reactivex.Completable.subscribe(Completable.java:2309)
        at io.reactivex.internal.operators.completable.CompletableOnErrorComplete.subscribeActual(CompletableOnErrorComplete.java:35)
        at io.reactivex.Completable.subscribe(Completable.java:2309)
        at io.reactivex.internal.operators.completable.CompletableSubscribeOn$SubscribeOnObserver.run(CompletableSubscribeOn.java:64)
        at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:38)
        at io.reactivex.internal.schedulers.ScheduledDirectTask.call(ScheduledDirectTask.java:26)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:301)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641)
        at java.lang.Thread.run(Thread.java:919)
     Caused by: AmplifyException {message=Failure performing sync query to AppSync., cause=AmplifyException {message=Could not retrieve the response body from the returned JSON, cause=java.io.IOException: Failed to retrieve Cognito User Pools token., recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.datastore.appsync.AppSyncClient.lambda$sync$1(AppSyncClient.java:97)
        at com.amplifyframework.datastore.appsync.-$$Lambda$AppSyncClient$suvEma05s-p8v-66iE4cswLKYac.accept(Unknown Source:4)
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onFailure(AppSyncGraphQLOperation.java:150)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:525)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
     Caused by: AmplifyException {message=Could not retrieve the response body from the returned JSON, cause=java.io.IOException: Failed to retrieve Cognito User Pools token., recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.api.aws.AppSyncGraphQLOperation$OkHttpCallback.onFailure(AppSyncGraphQLOperation.java:150) 
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:525) 
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
     Caused by: java.io.IOException: Failed to retrieve Cognito User Pools token.
        at com.amplifyframework.api.aws.sigv4.AppSyncSigV4SignerInterceptor.intercept(AppSyncSigV4SignerInterceptor.java:210)
        at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.kt:109)
        at okhttp3.internal.connection.RealCall.getResponseWithInterceptorChain$okhttp(RealCall.kt:201)
        at okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:517)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1167) 
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:641) 
        at java.lang.Thread.run(Thread.java:919) 
    Caused by: AmplifyException {message=getTokens does not support retrieving tokens while signed-out, cause=null, recoverySuggestion=Sorry, we don't have a suggested fix for this error yet.}
        at com.amplifyframework.api.aws.sigv4.DefaultCognitoUserPoolsAuthProvider.fetchToken(DefaultCognitoUserPoolsAuthProvider.java:97)
        at com.amplifyframework.api.aws.sigv4.DefaultCognitoUserPoolsAuthProvider.getLatestAuthToken(DefaultCognitoUserPoolsAuthProvider.java:103)
        at com.amplifyframework.api.aws.sigv4.AppSyncSigV4SignerInterceptor.intercept(AppSyncSigV4SignerInterceptor.java:208)
        	... 6 more`

Which I believe will cause problems in the future

**Comment by jamesonwilliams** [comment]
@Zablas Based on the bottom of your stack trace, it looks like you haven't signed-in any user. It also looks like you've configured this API to use Cognito. As a result, the DataStore is not able to sync, since it trying to use Cognito authenticaton when talking to AppSync.

TLDR: did you complete a successful `Amplify.Auth.signIn(...)`?

**Comment by Zablas** [comment]
@jamesonwilliams Yes signing in fixes that. Once again, the fix for the original bug is a complete reinstall of the app so I'm closing this for now. Thanks for the communication!
